docs/hugo/content/contributing/generator-overview.md

```diff
@@ -0,0 +1,56 @@
+| `astbuilder`       | Intention revealing utility methods for creating the underlying Go abstract syntax tree we serialize as the last step of generation.                  |
+| `codegen`          | The core processing pipeline of the code generator                                                                                                    |
+| `codegen/pipeline` | Individual pipeline stages that are composed to form the code generator itself.                                                                       |
+| `test`             | Support methods to make writing tests easier                                                                                                          |
+
+## Object Model
+
+At the core, the code generator works with a rich semantic object model that captures the structure of the resources (and related types) we are generating.
+
+Underpinning this object model is the `astmodel` package.
+
+The interface `astmodel.Type` represents a specific data type. The most widely used implementations of `Type` fall into a few separate groups:
+
+* Simple Go types, including `PrimitiveType`, `ArrayType`, and `MapType`.
+* Complex types with internal structure, most notably `ObjectType` and `ResourceType`.
+* Wrapper types that provide additional semantics, including `OptionalType`, `ErroredType`, `FlaggedType` and `ValidatedType`.
+
+This list is not exhaustive; other implementations of `Type` are used within limited scopes. For example, `AllOfType` and `OneOfType` represent JSON Schema techniques for creating object definitions via composition.
+
+Usefully, there is also `TypeName` which is both a type in itself and an indirect reference to a type defined elsewhere.
+
+When a `Type` is given a `TypeName`, it becomes a `TypeDefinition` and can be emitted as the source code for a Go type definition.
+
+Both `ResourceType` and `ObjectType` act as containers, each implementing `PropertyContainer`, `FunctionContainer`, and `TestCaseContainer`. These do pretty much what you'd expect from the names, though the implementations differ between `ResourceType` and `ObjectType`. For example, where an `ObjectType` implements `PropertyContainer` by providing support for an arbitrary set of properties, `ResourceType` has only `Spec` and `Status`.
+
+Most implementations of `astmodel.Function` are found in the **functions** package. New function implementations should go here; existing implementations are slowly being relocated.
+
+All implementations of `astmodel.TestCase` are found in the **testcases** package. New test case implementations should go here.
+
+## Generator Pipeline
+
+The code generator itself, found in the package `codegen`, is structured as a pipeline, composed of a series of stages that transform our object model incrementally. All the pipeline stages are found in the subpackage `codegen/pipeline`. 
```

Comment:
The mental model that I have (YMMV on if it's useful for you) is that `astmodel.Types` is a big graph of object and resource types, with the resources at the root and the types that the resources are built out of (spec types and the types that built the specs, status and the types that built the status and so on) as intermediate and leaf nodes.

The pipeline is all about making modifications to this graph by adding new types, changing existing types, etc. So each pipeline stage is a transformation of that graph.

It might be worth digging a bit into "how to think about the pipeline" - saying that they transform the object model is true but it doesn't necessarily explain what exactly that means.

Reply:
I've added some notes.